aggregator/src/aggregator/key_rotator.rs

```diff
@@ -0,0 +1,922 @@
+            }) {
+                ops.push(HpkeOp::Create(
+                    *ciphersuite,
+                    HpkeKeyState::Pending,
+                    "bootstrapping new configuration",
+                ))
+            }
+        }
+
+        // Partition existing keypairs by ciphersuite and state. The keypairs are sorted in
+        // descending order by `last_state_change_at`, i.e. the latest is first.
+        let mut by_ciphersuite: HashMap<
+            HpkeCiphersuite,
+            HashMap<HpkeKeyState, VecDeque<&GlobalHpkeKeypair>>,
+        > = HashMap::new();
+        for (_, keypair) in self.keypairs.iter() {
+            by_ciphersuite
+                .entry(keypair.ciphersuite())
+                .and_modify(|by_state| {
+                    by_state
+                        .entry(*keypair.state())
+                        .and_modify(|entry| {
+                            entry.insert(
+                                entry.partition_point(|&candidate| {
+                                    candidate.last_state_change_at()
+                                        > keypair.last_state_change_at()
+                                }),
+                                keypair,
+                            )
+                        })
+                        .or_insert_with(|| VecDeque::from([keypair]));
+                })
+                .or_insert_with(|| HashMap::from([(*keypair.state(), VecDeque::from([keypair]))]));
+        }
+
+        for (ciphersuite, mut by_state) in by_ciphersuite {
+            let active_keypairs = by_state.remove(&HpkeKeyState::Active).unwrap_or_default();
+            let pending_keypairs = by_state.remove(&HpkeKeyState::Pending).unwrap_or_default();
+            let expired_keypairs = by_state.remove(&HpkeKeyState::Expired).unwrap_or_default();
+
+            if self.config.ciphersuites.contains(&ciphersuite) {
+                let latest_pending_key = pending_keypairs.front();
+                let latest_active_key = active_keypairs.front();
+
+                if let Some(latest_pending_key) = latest_pending_key {
+                    if duration_since(&self.clock, latest_pending_key.last_state_change_at())
+                        > self.config.pending_duration
+                    {
+                        ops.push(HpkeOp::Update(
+                            *latest_pending_key.id(),
+                            HpkeKeyState::Active,
+                            "ready for promotion",
+                        ));
+
+                        if let Some(latest_active_key) = latest_active_key {
+                            ops.push(HpkeOp::Update(
+                                *latest_active_key.id(),
+                                HpkeKeyState::Expired,
+                                "pending key ready",
+                            ));
+                        }
+                    }
+                } else if latest_pending_key.is_none()
```

Comment:
nit: the `latest_pending_key.is_none()` check is extraneous, as we can only reach this block if the `if let Some(latest_pending_key) = latest_pending_key` above fails.

Reply:
(If it's written this way for clarity, replace with a comment?)